### DISCUSSION TITLE
Merging sync and async validation?

### DISCUSSION THREAD

**Discussion by juliendelort** [TanStack/form#488] [Ideas] [answered]
I don't know the motivation behind it but I wonder if we actually need to separate sync and async validation? For example, instead of having onChange and onChangeAsync, it could be simpler if we just had onChange that would accept a function that either returns a promise (async) or directly a validation error (sync)?

**Comment by tannerlinsley** [comment, marked as answer]
They’re separate so we can perform denouncing on one, but not the other. If you mix them, your execution control between the two is gone.

**Comment by juliendelort** [comment]
Makes sense, thank you Tanner!

**Comment by kinthaiofficial** [comment]
The sync vs async validation design in TanStack Form (and form libraries generally) is genuinely hard to get right. A few patterns:

**The core problem:**
Sync validation (type checking, required fields) should run on every keystroke. Async validation (server-side uniqueness checks, API lookups) should debounce to avoid hammering the server. Treating them identically leads to either too many API calls or sluggish local validation.

**The three-tier validation model that works:**

**1. Field-level sync validation (on change, instant):**
```typescript
<form.Field
  name="email"
  validators={{
    onChange: ({ value }) => {
      if (!value.includes('@')) return 'Must be a valid email';
    }
  }}
/>
```

**2. Field-level async validation (on blur, debounced):**
```typescript
<form.Field
  name="username"
  validators={{
    onBlurAsync: async ({ value }) => {
      const { isAvailable } = await checkUsername(value);
      if (!isAvailable) return 'Username taken';
    },
    onBlurAsyncDebounceMs: 500,
  }}
/>
```

**3. Form-level validation (on submit):**
```typescript
const form = useForm({
  onSubmit: async ({ value }) => {
    // Final server validation before submitting
  }
});
```

**The "merging" approach question:**
Your discussion title asks about merging sync and async validation. One clean approach: both return `string | undefined` (error or no error), and the form library merges: any non-undefined error wins. Async results override sync results for the same field once the async completes.

**What would actually be useful:** A validator that can short-circuit — if sync validation fails, skip the async call for that field.

**Reply by crutchcorn** [reply]
The short circuit logic you're describing already occurs
